README.md
# C2 Alignment Calculator

The Alignment Calculator is a C2 VALAWAI component that computes the alignment
if a set of norms or (*normative system*) with respect to a value of interest
using simulation and sampling.

The Alignment calculator works with an arbitrary *model*, which in practise is
implemented as a class. We denote the *state* of the model by $\mathbf{s}$. The
state $\mathbf{s}$ transitions at every time step according to the set of norms
$N=\{n_i\}$ in place. We denote a norm-regulated transition by $\mathbf{s}
\xrightarrow{N} \mathbf{s'}$. A succession of transitions $[\mathbf{s}_0
\xrightarrow{N} \mathbf{s}_1, ..., \mathbf{s}_{f-1} \xrightarrow{N}
\mathbf{s}_{f}]$ constitutes a *path*. The last state in a path $\mathbf{s}_f$
is an *outcome*.

To evaluate whether some set of norms $N$ is aligned with respect to a value $v$
or not, value $v$ has to be *grounded* in the context of the model we are
working with. This is done through the *value semantics function* $f_v:
\mathcal{S} \rightarrow [-1, 1]$, which maps the state of the model (i.e. the
object) to a number. The value semantics function evaluates the outcomes that
the norms $N$ steer the system towards. When this evaluation is positive, we
state that the norms $N$ *are aligned with respect to value $v$*. Formally:

$$
    \mathsf{Algn}_{N,v} = \mathbb{E}_N \left[f_v(\mathbf{S}_f)\right]
$$

The above states that the alignment $\mathsf{Algn}_{N,v}$ of the norms in $N$
with respect to value $v$ is computed as the expectation of the value semantics
function applied to the set of outcomes $\mathbf{S}_f$ under the implementation
of norms $N$ in the system. In practice, this expectation is computed by
simulating the evolution of the model under $N$ for a fixed number of time steps
(the *path length*), sampling their outcomes and evaluating them.

# Summary

 - Type: C2
 - Name: Alignment Calculator
 - Version: 1.0.0 (September 27, 2023)
 - API: [1.0.0 (February 3, 2023)](https://editor-next.swagger.io/?url=https://raw.githubusercontent.com/VALAWAI/C2_alignment_calculator/main/component-api.yml)
 - VALAWAI API: [0.1.0 (September 18, 2023)](https://editor-next.swagger.io/?url=https://raw.githubusercontent.com/VALAWAI/MOV/main/valawai-api.yml)
 - Developed by: IIIA-CSIC
 - License: [MIT](LICENSE)

# Usage

To understand how to set up and use this component, the reader is directed to
the `example.py` script available in this repository. There, you can find first
the `TaxModel` class. This class defined the model where norms are implemented
and where values are evaluated. The `TaxModel` class has a `step(norms)` method,
that takes in a `norms` dictionary and advances the state of the model by one
time step. The `norms` dictionary has a structure analogous to the following:

```python
{
    'pay': {
        'r0': 0.1,
        'r1': 0.2,
        'r2': 0.3,
        'r3': 0.4,
        'r4': 0.5
    },
    'payback': {
        'r0': 0.2,
        'r1': 0.2,
        'r2': 0.2,
        'r3': 0.2,
        'r4': 0.2
    }
}
```

The above norms dictionary indicates that there are two norms: `pay` (related to
the rates of tax payment) and `payback` (related to the rates of tax refunds).
Each norm is tied to a set of parameters, stating what is the rate of tax
payment or refund according to the wealth group of each agent in the model.

Next, the `example.py` script has two *value sematics functions* defined. These
are `ratio_wealth_value` and `gini_index_value`. Implemented value semantics
functions always take as input an instance of the model and return a float.

The Alignment Calculator component is implemented as a
[Flask](https://flask.palletsprojects.com/en/2.3.x/) application. To initialize
it, use the `create_app` function:

```python
from app import create_app

app = create_app(
    YourModel,                      # your model class
    [...],                          # your model initialization arguments
    {...},                          # your model initialization keyword arguments
    norms,                          # your norms dictionary
    your_value_semantics_funcion,   # your value semantics function
    path_length=10,                 # change if needed, default is 10
    path_sample=500                 # change if needed, default is 500
)
```

To develop your own model to be deployed in a component, the `template.py`
script is provided as a blueprint.

This component communicates through the following HTTP requests:

* Data messages:

    - GET `/algn`

* Control messages:

    - PATCH `/norms` changes the normative system
    - PATCH `/path_length` changes the length of the paths used to sample
      outcomes
    - PATH `/path_sample` changes the number of paths sampled to compute the
      alignment

# Deployment

Clone this repository and develop your model and value semantics functions
following the blueprint in `template.py`:

```bash
$ git clone https://github.com/VALAWAI/C2_alignment_calculator.git
```

Build your Docker image in your directory of the component repository:

```bash
$ cd /path/to/c2_alignment_calculator
$ docker build -t c2_alignment_calculator .
```

Run a Docker container with your C2 Alignment Calculator:

```bash
$ docker run --rm -d \
  --network valawai \
  --name c2_alignment_calculator \
  --mount type=bind,src="$(pwd)",target=/app \
  -p 5432:5000 \
  -e MODEL=my_model \
  c2_alignment_calculator
```

The environment variable `MODEL` refers to the script where you have defined
your model (do not include the .py extension).

Once the container is up and running, use `curl` to communicate with the
component:

```bash
$ curl http://localhost:5432/algn
{
  "algn": 0.9048607793661254
}
```

```bash
$ curl -X PATCH http://localhost:5432/path_sample -d 200
{}
```


CHANGELOG.md
# C2_alignment_calculator
 
## Version 1.0.0 (October 3, 2023)

- Initial component version.


example.py
import argparse
from mesa import Agent, Model
from mesa.time import BaseScheduler
from random import seed, uniform

from app import create_app

parser = argparse.ArgumentParser()
parser.add_argument('port', type=int)

seed(100)

class TaxModel(Model):
    def __init__(self) -> None:
        super().__init__()
        self.num_agents = 200
        self.num_segments = 5
        self.schedule = BaseScheduler(self)
        for i in range(self.num_agents):
            a = Agent(i, self)
            a.wealth = uniform(0, 100)
            self.schedule.add(a)
        self.sorted_agents = sorted(self.schedule.agents, key=lambda a: a.wealth)
        self.ags_per_seg = self.num_agents//self.num_segments
        self.segments = [i//self.ags_per_seg for i in range(self.num_agents)]

    def step(self, norms) -> None:
        common_fund = 0.
        for i in range(self.num_agents):
            a = self.sorted_agents[i]
            seg = self.segments[i]
            pay = a.wealth * norms['pay'][f"r{seg}"]
            if pay > a.wealth:
                pay = a.wealth
            a.wealth -= pay
            common_fund += pay
        common_fund *= 1.05
        segs_payback = [
            common_fund * norms['payback'][f"r{seg}"] / self.ags_per_seg
            for seg in range(self.num_segments)
        ]
        for i in range(self.num_agents):
            a = self.sorted_agents[i]
            seg = self.segments[i]
            a.wealth += segs_payback[seg]
        self.sorted_agents = sorted(self.schedule.agents, key=lambda a: a.wealth)


def ratio_wealth_value(mdl: TaxModel) -> float:
    """Semantics function for value equality based on the ratio between the
    maximum and the minimum wealth.

    Parameters
    ----------
    mdl : TaxModel

    Returns
    -------
    float
    """
    a = mdl.sorted_agents[99].wealth
    b = mdl.sorted_agents[199].wealth
    return a/b


def gini_index_value(mdl: TaxModel) -> float:
    """Semantics function for value equality based on the Gini Index.

    Parameters
    ----------
    mdl : TaxModel

    Returns
    -------
    float
    """
    sorted_wealth = [ag.wealth for ag in mdl.sorted_agents]
    N = mdl.num_agents
    B = sum(xi * (N - i) for i, xi in enumerate(sorted_wealth)) 
    GI =  1 + (1 / N) - 2 * B / (N * sum(sorted_wealth))
    return 1 - 2*GI

        
if __name__ == '__main__':

    normative_system = {
        'pay': {f'r{i}': 0.1+i/10 for i in range(5)},
        'payback': {f'r{i}': 0.2 for i in range(5)}
    }

    app = create_app(
        TaxModel,
        [],
        {},
        normative_system,
        gini_index_value,
        path_length=5,
        path_sample=100
    )

    args = parser.parse_args()
    app.run(debug=True, host='0.0.0.0', port=args.port)


template.py
import argparse
from app import create_app

from typing import Any

parser = argparse.ArgumentParser()
parser.add_argument('port', type=int)

class YourModel:
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        """A model whose evolution is determined by a set of norms."""
        super().__init__(*args, **kwargs)

    def step(
        self,
        norms: dict[str, dict[str, Any]]
    ) -> None:
        """The model evolves according to the set of norms in place.

        By having the norms as an input to the model's ``step()`` method, it is
        possible to model changes in the implemented normative system while the
        model is still running.

        Parameters
        ----------
        norms : Dict[Any, Dict[Any, Any]]
            A map of the norms governing the evolution of the model at that
            step. This map is provided as a dictionary from norm identifiers
            (keys) to a dictionary mapping each of the norms' parameters to
            their values.

        Raises
        ------
        NotImplementedError
        """
        raise NotImplementedError
    

def your_value_semantics_funcion(mdl: YourModel) -> float:
    """Value semantics function.

    Parameters
    ----------
    mdl : YourModel

    Returns
    -------
    float
        The degree of respect for the value at the current state of the model.

    Raises
    ------
    NotImplementedError
    """
    raise NotImplementedError


if __name__ == '__main__':

    norms = {
        'n1': {'p11': ...}
    }

    app = create_app(
        YourModel,                      # your model class
        [...],                          # your model initialization arguments
        {...},                          # your model initialization keyword arguments
        norms,                          # your norms dictionary
        your_value_semantics_funcion,   # your value semantics function
        # path_length=10,               # change if needed, default is 10
        # path_sample=500               # change if needed, default is 500
    )

    args = parser.parse_args()
    app.run(debug=True, host="0.0.0.0", port=args.port)


alignment_calculator.py
from flask import Flask, Request, request
from pathos.multiprocessing import cpu_count, ProcessPool
from traceback import format_exc

from typing import Any, Callable, Type


def evaluate_path(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    norms: dict[str, dict[str, Any]],
    value: Callable[[object], float],
    path_length: int
) -> float:
    """Evaluate the outcome of a path in terms of a value.

    Parameters
    ----------
    model_cls : Type[...]
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initialization keyword arguments.
    norms : dict[str, dict[str, Any]]
        The set of norms governing the evolution of the model.
    value : Callable[[object], float]
        The value semantics function that evaluates the final state of a path.
    path_length : int
        The number of steps in the path to evaluate.

    Returns
    -------
    float
    """
    mdl = model_cls(*model_args, **model_kwargs)
    for _ in range(path_length):
        mdl.step(norms)
    return value(mdl)


def alignment(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    norms: dict[str, dict[str, Any]],
    value: Callable[[object], float],
    path_length: int,
    path_sample: int,
    pool: ProcessPool
) -> float:
    """Compute the alignment from a sample of paths.

    This function uses a ``pathos.multiprocessing.ProcessPool`` already
    initialized to speed up the sampling.

    Parameters
    ----------
    model_cls : Type[Model]
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initialization keyword arguments.
    norms : dict[str, dict[str, Any]]
        The set of norms governing the evolution of the model.
    value : Callable[[object], float]
        The value semantics function that evaluates the final state of a path.
    path_length : int
        The number of steps in the path to evaluate.
    path_sample : int
        The number of paths to sample.
    pool : ProcessPool
        A ``pathos.multiprocessing.ProcessPool`` to parallelize sampling.

    Returns
    -------
    float
    """
    args = [
        [model_cls] * path_sample,
        [model_args] * path_sample,
        [model_kwargs] * path_sample,
        [norms] * path_sample,
        [value] * path_sample,
        [path_length] * path_sample
    ]
    pool.restart()
    algn = 0.
    for res in pool.map(evaluate_path, *args):
        algn += res
    pool.close()
    pool.terminate()
    return algn / path_sample
    

def create_app(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    value: Callable[[object], float]
) -> Flask:
    """Create a Flask app that computes the alignment with respect to a value.

    This C2 component of the VALAWAI architecture computes the alignment of a
    normative system with respect to a value (or an aggregation of values) [1]_.

    Parameters
    ----------
    model_cls : Type[...]
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initilization keyword arguments.
    value : Callable[[object], float]
        The value with respect to which the alignment is computed. It is passed
        as a function taking as input a model instance and returning the
        evaluation of the value semantics function given the state of the model.

    Returns
    -------
    Flask
        A Flask application that can process GET /algn requests.

    References
    ----------
    .. [1] Montes, N., & Sierra, C. (2022). Synthesis and properties of
        optimally value-aligned normative systems. Journal of Artificial
        Intelligence Research, 74, 1739–1774. https://doi.org/10.1613/jair.1.
        13487
    """
    __EXPECTED_TYPES = {
        'normative_system': dict, 'path_length': int, 'path_sample': int
    }
    
    app = Flask(__name__)

    def __check_request(request: Request):
        if not request.is_json:
            return {"error": "Request must be JSON"}, 415
        input_data = request.get_json()
        if not isinstance(input_data, dict):
            return {"error": f"Params must be passed as a dict"}, 400
        return input_data


    @app.get('/algn')
    def get_algn():
        input_data = __check_request(request)

        __NEED_KEYS = ['normative_system']
        __OPT_KEYS = ['path_length', 'path_sample']
        __DEFAULTS = {'path_length': 10, 'path_sample': 100}

        # check that the input data has all the necessary keys and they are the
        # correct type
        for k in __NEED_KEYS:
            if not k in input_data.keys():
                return {"error": f"missing necessary param {k}"}, 400
        for k in input_data.keys():
            if not isinstance(input_data[k], __EXPECTED_TYPES[k]):
                return {"error": f"type of param {k} must be {__EXPECTED_TYPES[k]}"}, 400

        # get optional arguments path_length and path_sample
        kwargs = {}
        for k in __OPT_KEYS:
            try:
                kwargs[k] = input_data[k]
            except KeyError:
                kwargs[k] = __DEFAULTS[k]

        # compute and return
        try:
            # prepare pool for parallelization
            num_nodes = cpu_count()
            if kwargs['path_sample'] <= num_nodes:
                num_nodes = kwargs['path_sample']
            pool = ProcessPool(nodes=num_nodes)

            algn = alignment(
                model_cls,
                model_args,
                model_kwargs,
                input_data['normative_system'],
                value,
                kwargs['path_length'],
                kwargs['path_sample'],
                pool
            )
            return {'algn': algn}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    return app


app.py
# Copyright 2023 IIIA-CSIC
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License version 3
# along with this program; if not, see <http://www.gnu.org/licenses/>.

from dataclasses import dataclass
from flask import Flask, Request, request
from traceback import format_exc
from valalgn.sampling import alignment

from typing import Any, Callable, Type

@dataclass
class AlignmentCalculator:
    model_cls: Type
    model_args: list[Any]
    model_kwargs: dict[str, Any]
    norms: dict[str, dict[str, Any]]
    value: Callable[[object], float]
    path_length: int = 10
    path_sample: int = 500

    def compute_alignment(self) -> float:
        return alignment(
            self.model_cls,
            self.model_args,
            self.model_kwargs,
            self.norms,
            self.value,
            self.path_length,
            self.path_sample
        )

def create_app(
    model_cls: Type,
    model_args: list[Any],
    model_kwargs: dict[str, Any],
    norms: dict[str, dict[str, Any]],
    value: Callable[[object], float],
    path_length: int = 10,
    path_sample: int = 500
) -> Flask:
    """Create a Flask app that computes the alignment with respect to a value.

    This C2 component of the VALAWAI architecture computes the alignment of a
    normative system with respect to a value (or an aggregation of values) [1]_.

    Parameters
    ----------
    model_cls : Type
        Class of the model.
    model_args : list[Any]
        Model initilization arguments.
    model_kwargs : dict[str, Any]
        Model initilization keyword arguments.
    norms : dict[str, dict[str, Any]]
        The set of norms governing the evolution of the model.
    value : Callable[[object], float]
        The value with respect to which the alignment is computed. It is passed
        as a function taking as input a model instance and returning the
        evaluation of the value semantics function given the state of the model.
    path_length : int, optional
        The length of the paths to evaluate the alignment, by default 10.
    path_sample : int, optional
        The sample size of the paths to evaluate the alignment, by default 500.

    Returns
    -------
    Flask
        A Flask application that can process GET /algn requests.

    References
    ----------
    .. [1] Montes, N., & Sierra, C. (2022). Synthesis and properties of
        optimally value-aligned normative systems. Journal of Artificial
        Intelligence Research, 74, 1739–1774. https://doi.org/10.1613/jair.1.
        13487
    """
    app = Flask(__name__)
    
    calculator = AlignmentCalculator(
        model_cls, model_args, model_kwargs, norms, value, path_length, path_sample
    )

    def __check_request(request: Request):
        if not request.is_json:
            return {"error": "Request must be JSON"}, 415
        input_data = request.get_json()
        if not isinstance(input_data, dict):
            return {"error": f"Params must be passed as a dict"}, 400
        return input_data

    @app.get('/algn')
    def get_algn():
        try:
            algn = calculator.compute_alignment()
            return {'algn': algn}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/norms')
    def patch_norms():
        input_data = __check_request(request)
        try:
            for k, v in input_data.items():
                calculator.norms[k].update(v)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/path_length')
    def patch_path_length():
        input_data = request.get_data()
        try:
            calculator.path_length = int(input_data)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    @app.patch('/path_sample')
    def patch_path_sample():
        input_data = request.get_data()
        try:
            calculator.path_sample = int(input_data)
            return {}, 200
        except Exception:
            return {"error": format_exc()}, 400
        
    return app
